// @ark-ui/react/tooltip/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const tooltip = require('./tooltip.cjs');
const tooltipArrow = require('./tooltip-arrow.cjs');
const tooltipArrowTip = require('./tooltip-arrow-tip.cjs');
const tooltipContent = require('./tooltip-content.cjs');
const tooltipContext = require('./tooltip-context.cjs');
const tooltipPositioner = require('./tooltip-positioner.cjs');
const tooltipTrigger = require('./tooltip-trigger.cjs');
const tooltip$1 = require('@zag-js/tooltip');



exports.Tooltip = tooltip.Tooltip;
exports.TooltipArrow = tooltipArrow.TooltipArrow;
exports.TooltipArrowTip = tooltipArrowTip.TooltipArrowTip;
exports.TooltipContent = tooltipContent.TooltipContent;
exports.useTooltipContext = tooltipContext.useTooltipContext;
exports.TooltipPositioner = tooltipPositioner.TooltipPositioner;
exports.TooltipTrigger = tooltipTrigger.TooltipTrigger;
Object.defineProperty(exports, 'tooltipAnatomy', {
	enumerable: true,
	get: () => tooltip$1.anatomy
});
