// @ark-ui/react/segment-group/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const segment = require('./segment.cjs');
const segmentContext = require('./segment-context.cjs');
const segmentControl = require('./segment-control.cjs');
const segmentGroup = require('./segment-group.cjs');
const segmentGroupContext = require('./segment-group-context.cjs');
const segmentGroupLabel = require('./segment-group-label.cjs');
const segmentGroup_anatomy = require('./segment-group.anatomy.cjs');
const segmentIndicator = require('./segment-indicator.cjs');
const segmentInput = require('./segment-input.cjs');
const segmentLabel = require('./segment-label.cjs');



exports.Segment = segment.Segment;
exports.useSegmentContext = segmentContext.useSegmentContext;
exports.SegmentControl = segmentControl.SegmentControl;
exports.SegmentGroup = segmentGroup.SegmentGroup;
exports.useSegmentGroupContext = segmentGroupContext.useSegmentGroupContext;
exports.SegmentGroupLabel = segmentGroupLabel.SegmentGroupLabel;
exports.segmentGroupAnatomy = segmentGroup_anatomy.segmentGroupAnatomy;
exports.SegmentIndicator = segmentIndicator.SegmentIndicator;
exports.SegmentInput = segmentInput.SegmentInput;
exports.SegmentLabel = segmentLabel.SegmentLabel;
