// @ark-ui/react/number-input/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const numberInput = require('./number-input.cjs');
const numberInputContext = require('./number-input-context.cjs');
const numberInputControl = require('./number-input-control.cjs');
const numberInputDecrementTrigger = require('./number-input-decrement-trigger.cjs');
const numberInputField = require('./number-input-field.cjs');
const numberInputIncrementTrigger = require('./number-input-increment-trigger.cjs');
const numberInputLabel = require('./number-input-label.cjs');
const numberInputScrubber = require('./number-input-scrubber.cjs');
const numberInput$1 = require('@zag-js/number-input');



exports.NumberInput = numberInput.NumberInput;
exports.useNumberInputContext = numberInputContext.useNumberInputContext;
exports.NumberInputControl = numberInputControl.NumberInputControl;
exports.NumberInputDecrementTrigger = numberInputDecrementTrigger.NumberInputDecrementTrigger;
exports.NumberInputField = numberInputField.NumberInputField;
exports.NumberInputIncrementTrigger = numberInputIncrementTrigger.NumberInputIncrementTrigger;
exports.NumberInputLabel = numberInputLabel.NumberInputLabel;
exports.NumberInputScrubber = numberInputScrubber.NumberInputScrubber;
Object.defineProperty(exports, 'numberInputAnatomy', {
  enumerable: true,
  get: () => numberInput$1.anatomy
});
