// @zag-js/checkbox/dist/index.mjs
// src/checkbox.anatomy.ts
import { createAnatomy } from "@zag-js/anatomy";
var anatomy = createAnatomy("checkbox").parts("root", "input", "label", "control");
var parts = anatomy.build();

// src/checkbox.connect.ts
import { dataAttr } from "@zag-js/dom-query";
import { visuallyHiddenStyle } from "@zag-js/visually-hidden";

// src/checkbox.dom.ts
import { createScope } from "@zag-js/dom-query";
var dom = createScope({
  getRootId: (ctx) => ctx.ids?.root ?? `checkbox:${ctx.id}`,
  getLabelId: (ctx) => ctx.ids?.label ?? `checkbox:${ctx.id}:label`,
  getControlId: (ctx) => ctx.ids?.control ?? `checkbox:${ctx.id}:control`,
  getInputId: (ctx) => ctx.ids?.input ?? `checkbox:${ctx.id}:input`,
  getRootEl: (ctx) => dom.queryById(ctx, dom.getRootId(ctx)),
  getInputEl: (ctx) => dom.queryById(ctx, dom.getInputId(ctx))
});

// src/checkbox.connect.ts
function connect(state, send, normalize) {
  const isDisabled = state.context.disabled;
  const isFocused = !isDisabled && state.context.focused;
  const isChecked = state.context.isChecked;
  const isIndeterminate2 = state.context.isIndeterminate;
  const dataAttrs = {
    "data-active": dataAttr(state.context.active),
    "data-focus": dataAttr(isFocused),
    "data-hover": dataAttr(state.context.hovered),
    "data-disabled": dataAttr(isDisabled),
    "data-state": isIndeterminate2 ? "indeterminate" : state.context.checked ? "checked" : "unchecked",
    "data-invalid": dataAttr(state.context.invalid)
  };
  return {
    isChecked,
    isDisabled,
    isIndeterminate: isIndeterminate2,
    isFocused,
    checkedState: state.context.checked,
    setChecked(checked) {
      send({ type: "CHECKED.SET", checked });
    },
    toggleChecked() {
      send({ type: "CHECKED.TOGGLE", checked: isChecked });
    },
    rootProps: normalize.label({
      ...parts.root.attrs,
      ...dataAttrs,
      id: dom.getRootId(state.context),
      htmlFor: dom.getInputId(state.context),
      onPointerMove() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { hovered: true } });
      },
      onPointerLeave() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { hovered: false } });
      },
      onPointerDown(event) {
        if (isDisabled)
          return;
        if (isFocused && event.pointerType === "mouse") {
          event.preventDefault();
        }
        send({ type: "CONTEXT.SET", context: { active: true } });
      },
      onPointerUp() {
        if (isDisabled)
          return;
        send({ type: "CONTEXT.SET", context: { active: false } });
      },
      onClick(event) {
        if (event.target === dom.getInputEl(state.context)) {
          event.stopPropagation();
        }
      }
    }),
    labelProps: normalize.element({
      ...parts.label.attrs,
      ...dataAttrs,
      id: dom.getLabelId(state.context)
    }),
    controlProps: normalize.element({
      ...parts.control.attrs,
      ...dataAttrs,
      id: dom.getControlId(state.context),
      "aria-hidden": true
    }),
    inputProps: normalize.input({
      ...parts.input.attrs,
      id: dom.getInputId(state.context),
      type: "checkbox",
      required: state.context.required,
      defaultChecked: isChecked,
      disabled: isDisabled,
      "data-disabled": dataAttr(isDisabled),
      "aria-labelledby": dom.getLabelId(state.context),
      "aria-invalid": state.context.invalid,
      name: state.context.name,
      form: state.context.form,
      value: state.context.value,
      style: visuallyHiddenStyle,
      onChange(event) {
        const checked = event.currentTarget.checked;
        send({ type: "CHECKED.SET", checked });
      },
      onBlur() {
        send({ type: "CONTEXT.SET", context: { focused: false } });
      },
      onFocus() {
        send({ type: "CONTEXT.SET", context: { focused: true } });
      },
      onKeyDown(event) {
        if (event.key === " ") {
          send({ type: "CONTEXT.SET", context: { active: true } });
        }
      },
      onKeyUp(event) {
        if (event.key === " ") {
          send({ type: "CONTEXT.SET", context: { active: false } });
        }
      }
    })
  };
}

// src/checkbox.machine.ts
import { createMachine } from "@zag-js/core";
import { dispatchInputCheckedEvent, trackFormControl } from "@zag-js/form-utils";
import { compact } from "@zag-js/utils";
function machine(userContext) {
  const ctx = compact(userContext);
  return createMachine(
    {
      id: "checkbox",
      initial: "ready",
      context: {
        checked: false,
        value: "on",
        ...ctx
      },
      watch: {
        disabled: "removeFocusIfNeeded",
        checked: ["invokeOnChange", "syncInputElement"]
      },
      activities: ["trackFormControlState"],
      on: {
        "CHECKED.TOGGLE": {
          actions: ["toggleChecked"]
        },
        "CHECKED.SET": {
          actions: ["setChecked"]
        },
        "CONTEXT.SET": {
          actions: ["setContext"]
        }
      },
      computed: {
        isIndeterminate: (ctx2) => isIndeterminate(ctx2.checked),
        isChecked: (ctx2) => ctx2.isIndeterminate ? false : !!ctx2.checked
      },
      states: {
        ready: {}
      }
    },
    {
      activities: {
        trackFormControlState(ctx2, _evt, { send, initialContext }) {
          return trackFormControl(dom.getInputEl(ctx2), {
            onFieldsetDisabled() {
              ctx2.disabled = true;
            },
            onFormReset() {
              send({ type: "CHECKED.SET", checked: !!initialContext.checked });
            }
          });
        }
      },
      actions: {
        invokeOnChange(ctx2) {
          ctx2.onChange?.({ checked: ctx2.checked });
        },
        setContext(ctx2, evt) {
          Object.assign(ctx2, evt.context);
        },
        syncInputElement(ctx2) {
          const inputEl = dom.getInputEl(ctx2);
          inputEl.checked = ctx2.isChecked;
          inputEl.indeterminate = ctx2.isIndeterminate;
        },
        dispatchCheckedEvent(ctx2, evt) {
          const inputEl = dom.getInputEl(ctx2);
          const checked = isIndeterminate(evt.checked) ? false : evt.checked;
          dispatchInputCheckedEvent(inputEl, { checked, bubbles: true });
        },
        removeFocusIfNeeded(ctx2) {
          if (ctx2.disabled && ctx2.focused) {
            ctx2.focused = false;
          }
        },
        setChecked(ctx2, evt) {
          ctx2.checked = evt.checked;
        },
        toggleChecked(ctx2) {
          ctx2.checked = isIndeterminate(ctx2.checked) ? true : !ctx2.checked;
        }
      }
    }
  );
}
function isIndeterminate(checked) {
  return checked === "indeterminate";
}
export {
  anatomy,
  connect,
  machine
};
