// @ark-ui/react/dialog/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const dialog = require('./dialog.cjs');
const dialogBackdrop = require('./dialog-backdrop.cjs');
const dialogCloseTrigger = require('./dialog-close-trigger.cjs');
const dialogContainer = require('./dialog-container.cjs');
const dialogContent = require('./dialog-content.cjs');
const dialogContext = require('./dialog-context.cjs');
const dialogDescription = require('./dialog-description.cjs');
const dialogTitle = require('./dialog-title.cjs');
const dialogTrigger = require('./dialog-trigger.cjs');
const dialog$1 = require('@zag-js/dialog');



exports.Dialog = dialog.Dialog;
exports.DialogBackdrop = dialogBackdrop.DialogBackdrop;
exports.DialogCloseTrigger = dialogCloseTrigger.DialogCloseTrigger;
exports.DialogContainer = dialogContainer.DialogContainer;
exports.DialogContent = dialogContent.DialogContent;
exports.useDialogContext = dialogContext.useDialogContext;
exports.DialogDescription = dialogDescription.DialogDescription;
exports.DialogTitle = dialogTitle.DialogTitle;
exports.DialogTrigger = dialogTrigger.DialogTrigger;
Object.defineProperty(exports, 'dialogAnatomy', {
	enumerable: true,
	get: () => dialog$1.anatomy
});
