// @ark-ui/react/editable/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const editable = require('./editable.cjs');
const editableArea = require('./editable-area.cjs');
const editableCancelTrigger = require('./editable-cancel-trigger.cjs');
const editableContext = require('./editable-context.cjs');
const editableControl = require('./editable-control.cjs');
const editableEditTrigger = require('./editable-edit-trigger.cjs');
const editableInput = require('./editable-input.cjs');
const editableLabel = require('./editable-label.cjs');
const editablePreview = require('./editable-preview.cjs');
const editableSubmitTrigger = require('./editable-submit-trigger.cjs');
const editable$1 = require('@zag-js/editable');



exports.Editable = editable.Editable;
exports.EditableArea = editableArea.EditableArea;
exports.EditableCancelTrigger = editableCancelTrigger.EditableCancelTrigger;
exports.useEditableContext = editableContext.useEditableContext;
exports.EditableControl = editableControl.EditableControl;
exports.EditableEditTrigger = editableEditTrigger.EditableEditTrigger;
exports.EditableInput = editableInput.EditableInput;
exports.EditableLabel = editableLabel.EditableLabel;
exports.EditablePreview = editablePreview.EditablePreview;
exports.EditableSubmitTrigger = editableSubmitTrigger.EditableSubmitTrigger;
Object.defineProperty(exports, 'editableAnatomy', {
	enumerable: true,
	get: () => editable$1.anatomy
});
