// @ark-ui/react/tags-input/index.cjs
'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const tag = require('./tag.cjs');
const tagDeleteTrigger = require('./tag-delete-trigger.cjs');
const tagInput = require('./tag-input.cjs');
const tagsInput = require('./tags-input.cjs');
const tagsInputClearTrigger = require('./tags-input-clear-trigger.cjs');
const tagsInputContext = require('./tags-input-context.cjs');
const tagsInputControl = require('./tags-input-control.cjs');
const tagsInputField = require('./tags-input-field.cjs');
const tagsInputLabel = require('./tags-input-label.cjs');
const tagsInput$1 = require('@zag-js/tags-input');



exports.Tag = tag.Tag;
exports.TagDeleteTrigger = tagDeleteTrigger.TagDeleteTrigger;
exports.TagInput = tagInput.TagInput;
exports.TagsInput = tagsInput.TagsInput;
exports.TagsInputClearTrigger = tagsInputClearTrigger.TagsInputClearTrigger;
exports.useTagsInputContext = tagsInputContext.useTagsInputContext;
exports.TagsInputControl = tagsInputControl.TagsInputControl;
exports.TagsInputField = tagsInputField.TagsInputField;
exports.TagsInputLabel = tagsInputLabel.TagsInputLabel;
Object.defineProperty(exports, 'tagsInputAnatomy', {
	enumerable: true,
	get: () => tagsInput$1.anatomy
});
